File: x/agentplane/runner/harness_supervisor.c
#include <errno.h>
#include <signal.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <sys/prctl.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <unistd.h>

static volatile sig_atomic_t stop_signal;

static void request_stop(int signal_number) {
  stop_signal = signal_number;
}

static int install_handler(int signal_number) {
  struct sigaction action = {
      .sa_handler = request_stop,
  };
  sigemptyset(&action.sa_mask);
  return sigaction(signal_number, &action, NULL);
}

static int signal_group(pid_t leader, int signal_number, const char *description) {
  if (kill(-leader, signal_number) == -1 && errno != ESRCH) {
    perror(description);
    return -1;
  }
  return 0;
}

int main(int argc, char *argv[]) {
  if (argc < 4 || strcmp(argv[1], "--native-pid-fd") != 0) {
    fputs("harness supervisor requires --native-pid-fd FD and a command\n", stderr);
    return 64;
  }
  char *end = NULL;
  long descriptor = strtol(argv[2], &end, 10);
  if (*argv[2] == '\0' || *end != '\0' || descriptor < 0) {
    fputs("harness supervisor received an invalid native pid descriptor\n", stderr);
    return 64;
  }
  int native_pid_fd = (int)descriptor;
  if (install_handler(SIGTERM) == -1 || install_handler(SIGUSR1) == -1 || install_handler(SIGALRM) == -1) {
    perror("install harness supervisor signal handler");
    return 125;
  }

  pid_t parent = getppid();
  if (prctl(PR_SET_PDEATHSIG, SIGUSR1) == -1) {
    perror("set harness parent-death signal");
    return 125;
  }
  if (getppid() != parent) {
    stop_signal = SIGUSR1;
  }
  if (stop_signal != 0) {
    return 128 + stop_signal;
  }

  pid_t child = fork();
  if (child == -1) {
    perror("fork harness");
    return 125;
  }
  if (child == 0) {
    close(native_pid_fd);
    if (setpgid(0, 0) == -1) {
      perror("set harness process group");
      _exit(125);
    }
    execvp(argv[3], argv + 3);
    perror("exec harness");
    _exit(127);
  }
  if (setpgid(child, child) == -1 && errno != EACCES && errno != ESRCH) {
    perror("set harness child process group");
    kill(child, SIGKILL);
    waitpid(child, NULL, 0);
    return 125;
  }
  if (dprintf(native_pid_fd, "%ld\n", (long)child) < 0) {
    perror("report harness pid");
    kill(-child, SIGKILL);
    waitpid(child, NULL, 0);
    return 125;
  }
  close(native_pid_fd);

  int stop_sent = 0;
  if (stop_signal != 0) {
    // A parent-death signal can arrive after the child starts but before the first waitpid.
    // Consume it here rather than entering a wait that no later signal will interrupt.
    stop_signal = 0;
    stop_sent = 1;
    alarm(5);
    if (signal_group(child, SIGTERM, "stop harness process group") == -1) {
      return 125;
    }
  }
  int status;
  while (waitpid(child, &status, 0) == -1) {
    if (errno != EINTR) {
      perror("wait for harness");
      return 125;
    }
    if (stop_signal == 0) {
      continue;
    }
    // On runner death the harness's stdin is already closed. Give it the same SIGTERM it gets
    // during orderly shutdown so it can persist its native resume state; this supervisor retains
    // the state-owner descriptor until that child group actually exits. A replacement therefore
    // fails closed rather than overlapping a native process that has not stopped yet.
    if (stop_signal == SIGALRM) {
      stop_signal = 0;
      alarm(0);
      if (signal_group(child, SIGKILL, "force-stop harness process group") == -1) {
        return 125;
      }
      continue;
    }
    if (!stop_sent) {
      // On runner death the harness's stdin is already closed. Give it the same SIGTERM it gets
      // during orderly shutdown so it can persist its native resume state. If it cannot exit,
      // the kernel timer force-fences the entire process group rather than letting a successor
      // wait forever with no writer available.
      stop_signal = 0;
      stop_sent = 1;
      alarm(5);
      if (signal_group(child, SIGTERM, "stop harness process group") == -1) {
        return 125;
      }
    }
  }
  alarm(0);
  if (!stop_sent && stop_signal != 0) {
    // The leader can exit at the same instant its runner dies. Its descendants still need the
    // parent-death fence even though there is no leader left to wait for.
    stop_signal = 0;
    stop_sent = 1;
    if (signal_group(child, SIGTERM, "stop remaining harness process group") == -1) {
      return 125;
    }
  }
  if (stop_sent && kill(-child, 0) == 0) {
    // The leader may have exited while a tool ignored its graceful stop. Do not release the
    // inherited ownership descriptor until that remaining group is force-fenced.
    if (signal_group(child, SIGKILL, "force-stop remaining harness process group") == -1) {
      return 125;
    }
  } else if (stop_sent && errno != ESRCH) {
    perror("check remaining harness process group");
    return 125;
  }
  if (WIFEXITED(status)) {
    return WEXITSTATUS(status);
  }
  if (WIFSIGNALED(status)) {
    return 128 + WTERMSIG(status);
  }
  return 125;
}

File: x/agentplane/runner/harness_supervisor.rs
use std::env;
use std::ffi::OsString;
use std::fs::File;
use std::io::Write;
use std::os::fd::{FromRawFd, RawFd};
use std::os::unix::process::{CommandExt, ExitStatusExt};
use std::process::{self, Child, Command, ExitStatus};

use anyhow::{Context, Result};
use nix::errno::Errno;
use nix::fcntl::{FcntlArg, FdFlag, fcntl};
use nix::sys::prctl::set_pdeathsig;
use nix::sys::signal::{SigSet, SigmaskHow, Signal, killpg};
use nix::unistd::{Pid, alarm, getppid};

fn signal_group(leader: Pid, signal: Signal) -> Result<()> {
    match killpg(leader, signal) {
        Ok(()) | Err(Errno::ESRCH) => Ok(()),
        result => result.with_context(|| format!("send {signal} to harness group {leader}")),
    }
}

fn wait_for_harness(child: &mut Child, signals: &SigSet) -> Result<ExitStatus> {
    let leader = Pid::from_raw(child.id() as i32);
    let mut stopping = false;
    loop {
        match signals
            .wait()
            .context("wait for harness supervisor signal")?
        {
            Signal::SIGTERM | Signal::SIGUSR1 if !stopping => {
                stopping = true;
                // Give the harness time to persist resume state. The supervisor retains the
                // inherited state-owner descriptor throughout graceful shutdown and escalation.
                alarm::set(5);
                signal_group(leader, Signal::SIGTERM)?;
            }
            Signal::SIGALRM => signal_group(leader, Signal::SIGKILL)?,
            _ => {}
        }
        if let Some(status) = child.try_wait().context("wait for harness")? {
            alarm::cancel();
            // No tool may outlive its native leader: after we exit there is no supervisor left
            // to receive parent-death notification and release an orphan's inherited owner lock.
            signal_group(leader, Signal::SIGKILL)?;
            return Ok(status);
        }
    }
}

fn supervise(mut report: File, program: &OsString, args: &[OsString]) -> Result<i32> {
    let parent = getppid();
    let mut signals = SigSet::empty();
    for signal in [
        Signal::SIGCHLD,
        Signal::SIGTERM,
        Signal::SIGUSR1,
        Signal::SIGALRM,
    ] {
        signals.add(signal);
    }
    // Blocking before spawn makes both exit and stop notifications pending until sigwait
    // consumes them; there is no signal-handler/check/wait race or shared signal-handler state.
    let original_mask = signals.thread_swap_mask(SigmaskHow::SIG_BLOCK)?;
    set_pdeathsig(Signal::SIGUSR1).context("set harness parent-death signal")?;
    if getppid() != parent {
        return Ok(128 + Signal::SIGUSR1 as i32);
    }

    // Only the supervisor reports the PID. Other inherited descriptors (especially the state
    // owner's open file description) deliberately survive exec into the harness and its tools.
    fcntl(&report, FcntlArg::F_SETFD(FdFlag::FD_CLOEXEC))?;
    let mut command = Command::new(program);
    command.args(args).process_group(0);
    // SAFETY: the pre-exec callback only restores the signal mask with pthread_sigmask, an
    // async-signal-safe syscall. It does not allocate or acquire Rust locks after fork.
    unsafe {
        command.pre_exec(move || {
            original_mask
                .thread_set_mask()
                .map_err(std::io::Error::from)
        });
    }
    let mut child = command.spawn().context("start native harness")?;
    let outcome = (|| {
        writeln!(report, "{}", child.id()).context("report harness pid")?;
        drop(report);
        wait_for_harness(&mut child, &signals)
    })();
    if outcome.is_err() {
        // Reporting/waiting failures must not leave a native writer alive after we release our
        // ownership descriptor. The child also retains that descriptor until it exits.
        signal_group(Pid::from_raw(child.id() as i32), Signal::SIGKILL)?;
        child
            .wait()
            .context("reap harness after supervisor failure")?;
    }
    let status = outcome?;
    Ok(status
        .code()
        .unwrap_or_else(|| 128 + status.signal().expect("reaped harness has an exit status")))
}

fn main() {
    let args: Vec<_> = env::args_os().skip(1).collect();
    let [flag, descriptor, program, args @ ..] = args.as_slice() else {
        eprintln!("harness supervisor requires --native-pid-fd FD and a command");
        process::exit(64);
    };
    if flag != "--native-pid-fd" {
        eprintln!("harness supervisor requires --native-pid-fd FD and a command");
        process::exit(64);
    }
    let Some(descriptor) = descriptor
        .to_str()
        .and_then(|value| value.parse::<RawFd>().ok())
        .filter(|value| *value >= 0)
    else {
        eprintln!("harness supervisor received an invalid native pid descriptor");
        process::exit(64);
    };
    // SAFETY: HarnessProcess passes ownership of this open pipe descriptor across exec.
    let report = unsafe { File::from_raw_fd(descriptor) };
    match supervise(report, program, args) {
        Ok(code) => process::exit(code),
        Err(error) => {
            eprintln!("harness supervisor: {error:#}");
            process::exit(125);
        }
    }
}
